ocaml/alerts/certificate/certificate_check.ml

```diff
@@ -41,39 +41,38 @@ let generate_alert epoch (host, expiry) =
     (host, Some (message, alert))
 
 let execute rpc session existing_messages (host, alert) =
-  let host_uuid = XenAPI.Host.get_uuid rpc session host in
-  let messages_in_host =
-    List.filter
-      (fun (_, record) -> record.API.message_obj_uuid = host_uuid)
-      existing_messages
-  in
-  let is_outdated (ref, record) =
-    match alert with
-    | None ->
-        true
-    | Some (body, (name, priority)) ->
-        record.API.message_body <> body
-        || record.API.message_name <> name
-        || record.API.message_priority <> priority
-  in
   (* CA-342551: messages need to be deleted if the pending alert regard the
      same host and has newer, updated information.
      If the pending alert has the same metadata as the existing host message
      it must not be emmited and the existing message must be retained, this
      prevents changing the message UUID.
      In the case there are alerts regarding the host but no alert is pending
      they are not destroyed since no alert is automatically dismissed. *)
-  let outdated_messages = List.filter is_outdated messages_in_host in
-  let replacement_is_pending = List.exists is_outdated messages_in_host in
-  match (alert, replacement_is_pending) with
-  | Some (message, (alert, priority)), true ->
+  match alert with
+  | Some (message, (alert, priority)) ->
+      let host_uuid = XenAPI.Host.get_uuid rpc session host in
+      let messages_in_host =
+        List.filter
+          (fun (_, record) -> record.API.message_obj_uuid = host_uuid)
+          existing_messages
+      in
+      let is_outdated (ref, record) =
+        record.API.message_body <> message
+        || record.API.message_name <> alert
+        || record.API.message_priority <> priority
+      in
+      let outdated, current = List.partition is_outdated messages_in_host in
+
       List.iter
         (fun (self, _) -> XenAPI.Message.destroy rpc session self)
-        outdated_messages ;
-      ignore
-        (XenAPI.Message.create rpc session alert priority `Host host_uuid
-           message)
-  | _ ->
+        outdated ;
+      if current = [] then
```

Comment:
Why do we need to do something when `current = []`? Should this be `<>`?

Reply:
`current` being empty means that the new alert is different from any existing alerts so it should be emitted.

This may happen in two cases:
1. there are no certificate alerts in the system about the host: the new alert is generated
2. There is at least one alert in the system but these have information that is different from the new alert: they are deleted (they are present in `outdated`) and the new alert is generated.

Reply:
So the message we need either exists already (in current), or we need to create it. It seems to me that `current` should have at most one member because to be in current all messages have to agree on message, alert, and priority. Not sure it's worth acting on this if more than one message is here.